### PR TITLE
Added back CacheControl headers from PR #6265

### LINKED ISSUE #5764: rpc: http jsonrpc don't set cache-control in response
## Summary

Some of the RPC apis can be perfectly cached since the data is immutable, like `/block` `/block_results`, etc.
If we set proper cache-control response header, and setup with some http proxy cache server like [varnish](https://varnish-cache.org/), we can easily archive great performance of http apis.

## Proposal

____

#### For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by Bro-Chain**
Issue #5764 raised the feature request of having the RPC return Cache-Control headers when applicable. This resulted in PR #6265, which was later merged. But it was merged into master, so it never made it to release in 0.34.x. This is that patch again, but so it can actually be released.

---
Closes #5764 (again)

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: light/proxy/routes.go [hunk 1]
```diff
@@ -18,37 +18,37 @@ func RPCRoutes(c *lrpc.Client) map[string]*rpcserver.RPCFunc {
...
+		"genesis_chunked":      rpcserver.NewRPCFunc(makeGenesisChunkedFunc(c), "", true),
+		"block":                rpcserver.NewRPCFunc(makeBlockFunc(c), "height", true),
+		"block_by_hash":        rpcserver.NewRPCFunc(makeBlockByHashFunc(c), "hash", true),
+		"block_results":        rpcserver.NewRPCFunc(makeBlockResultsFunc(c), "height", true),
+		"commit":               rpcserver.NewRPCFunc(makeCommitFunc(c), "height", true),
+		"tx":                   rpcserver.NewRPCFunc(makeTxFunc(c), "hash,prove", false),
+		"tx_search":            rpcserver.NewRPCFunc(makeTxSearchFunc(c), "query,prove,page,per_page,order_by", false),
+		"block_search":         rpcserver.NewRPCFunc(makeBlockSearchFunc(c), "query,page,per_page,order_by", false),
+		"validators":           rpcserver.NewRPCFunc(makeValidatorsFunc(c), "height,page,per_page", true),
+		"dump_consensus_state": rpcserver.NewRPCFunc(makeDumpConsensusStateFunc(c), "", true),
```
Comment (sergio-mena) [inline light/proxy/routes.go:35]: _dump_consensus_state_ has _false_ in the changes to `rpc/core/routes.go` below

File: light/proxy/routes.go [hunk 2]
```diff
@@ -18,37 +18,37 @@ func RPCRoutes(c *lrpc.Client) map[string]*rpcserver.RPCFunc {
...
+		"status":               rpcserver.NewRPCFunc(makeStatusFunc(c), "", false),
+		"net_info":             rpcserver.NewRPCFunc(makeNetInfoFunc(c), "", false),
+		"blockchain":           rpcserver.NewRPCFunc(makeBlockchainInfoFunc(c), "minHeight,maxHeight", true),
+		"genesis":              rpcserver.NewRPCFunc(makeGenesisFunc(c), "", true),
+		"genesis_chunked":      rpcserver.NewRPCFunc(makeGenesisChunkedFunc(c), "", true),
+		"block":                rpcserver.NewRPCFunc(makeBlockFunc(c), "height", true),
+		"block_by_hash":        rpcserver.NewRPCFunc(makeBlockByHashFunc(c), "hash", true),
+		"block_results":        rpcserver.NewRPCFunc(makeBlockResultsFunc(c), "height", true),
+		"commit":               rpcserver.NewRPCFunc(makeCommitFunc(c), "height", true),
+		"tx":                   rpcserver.NewRPCFunc(makeTxFunc(c), "hash,prove", false),
```
[the remark this answers is not in the document]
Comment (Bro-Chain) [reply light/proxy/routes.go]: Corrected

File: light/proxy/routes.go (hunk 1 above)
Comment (Bro-Chain) [reply light/proxy/routes.go]: Corrected

**Comment by JayT106** [comment]
same in PR#9442

File: rpc/jsonrpc/server/http_json_handler.go [hunk 3]
```diff
@@ -55,6 +55,11 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
 			requests = []types.RPCRequest{request}
 		}
 
+		// Set the default response cache to true unless
+		// 1. Any RPC request rrror.
```
Comment (sergio-mena) [inline rpc/jsonrpc/server/http_json_handler.go:59]:
```suggestion
		// 1. Any RPC request error.
```

**Review by sergio-mena — DISMISSED** [review]
Thanks for the changes, I added one more minor comment

**Comment by sergio-mena** [comment]
Please add a changelog entry

File: rpc/jsonrpc/server/http_server.go [hunk 4]
```diff
@@ -129,6 +130,9 @@ func WriteRPCResponseHTTP(w http.ResponseWriter, res ...types.RPCResponse) error
 		return fmt.Errorf("json marshal: %w", err)
 	}
 	w.Header().Set("Content-Type", "application/json")
+	if c {
+		w.Header().Set("Cache-Control", "public, max-age=86400") // expired after one year
```
Comment (sergio-mena) [inline rpc/jsonrpc/server/http_server.go:134]: Isn't it now one day?

**Comment by sergio-mena** [comment]
> link to PR #9442

I took a look at #9442. My understanding is that #9442 is equivalent to this PR (but for `main`, whereas this is for `v0.34.x`). So I have the following questions:

* What about `v0.37.x`
* PR #9442 seems more advanced than this PR in that it uses extensible options (rather than an extra field), so **why are we creating two implementations of the same feature?**; could #9442 be backported #9442 to v0.34.x (and v0.37.x)? Is it that this PR is a *pure* backport, whereas #9442 is a re-implementation?

**Comment by thanethomson** [comment]
Supeseded by #9650.

**Pull request closed without merging**